### PR TITLE
move BalanceReportInfo to mojo type

### PR DESCRIPTION
**Pull request by cg505**
Resolves https://github.com/brave/brave-browser/issues/5126

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [x] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/services/bat_ledger/bat_ledger_impl.cc [hunk 1]
```diff
@@ -287,19 +287,19 @@ void BatLedgerImpl::OnTimer(uint32_t timer_id) {
...
-    out_reports[report.first] = report.second.ToJson();
+    out_reports[report.first] = ledger::BalanceReportInfo::New(report.second);
   }
-  std::move(callback).Run(out_reports);
+  std::move(callback).Run(std::move(out_reports));
 }
 
 void BatLedgerImpl::GetBalanceReport(int32_t month, int32_t year,
     GetBalanceReportCallback callback) {
   ledger::BalanceReportInfo info;
```
Comment (NejcZdovc) [inline components/services/bat_ledger/bat_ledger_impl.cc]: you can just do `auto info = ledger::BalanceReportInfo::New()`

File: components/services/bat_ledger/bat_ledger_impl.cc [hunk 2]
```diff
@@ -287,19 +287,19 @@ void BatLedgerImpl::OnTimer(uint32_t timer_id) {
 void BatLedgerImpl::GetAllBalanceReports(
     GetAllBalanceReportsCallback callback) {
   auto reports = ledger_->GetAllBalanceReports();
-  base::flat_map<std::string, std::string> out_reports;
+  base::flat_map<std::string, ledger::BalanceReportInfoPtr> out_reports;
```
Comment (NejcZdovc) [inline components/services/bat_ledger/bat_ledger_impl.cc]: could we just do here something like:

```
void BatLedgerImpl::GetAllBalanceReports(GetAllBalanceReportsCallback callback) {
  auto reports = ledger_->GetAllBalanceReports();
  auto out_reports = mojo::MapToFlatMap(reports):
  std::move(callback).Run(std::move(out_reports));
}
```

File: vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc [hunk 3]
```diff
@@ -634,13 +634,13 @@ BatPublishers::getAllBalanceReports() {
...
     ledger::BalanceReportInfo newReport;
     const braveledger_bat_helper::REPORT_BALANCE_ST oldReport = report.second;
-    newReport.opening_balance_ = oldReport.opening_balance_;
-    newReport.closing_balance_ = oldReport.closing_balance_;
-    newReport.grants_ = oldReport.grants_;
-    newReport.earning_from_ads_ = oldReport.earning_from_ads_;
-    newReport.auto_contribute_ = oldReport.auto_contribute_;
-    newReport.recurring_donation_ = oldReport.recurring_donation_;
-    newReport.one_time_donation_ = oldReport.one_time_donation_;
+    newReport.opening_balance = oldReport.opening_balance_;
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc]: rename `BatPublishers::getAllBalanceReports()` -> `BatPublishers::GetAllBalanceReports()`

change return `std::map<std::string, ledger::BalanceReportInfo>` -> `std::map<std::string, ledger::BalanceReportInfoPtr>` (this way you don't need to convert it to ptr in `bat_ledger_impl`

File: components/services/bat_ledger/bat_ledger_impl.cc (hunk 2 above)
Comment (cg505) [reply components/services/bat_ledger/bat_ledger_impl.cc]: Problem is that `GetAllBalanceReportsCallback` expects the value type to be `BalanceReportInfoPtr`, but `MapToFlatMap` will give a `base::flatt_map<std::string, ledger::BalanceReportInfo>`

Comment (cg505) [reply components/services/bat_ledger/bat_ledger_impl.cc]: `MapToFlatMap` doesn't want to work on this type, but you can work around it.

**Comment by cg505** [comment]
Need to investigate the CI failures.

**Pull request merged**